### DISCUSSION TITLE
What "case" to choose

### DISCUSSION THREAD

**Discussion by kellijohnson-NOAA** [pfmc-assessments/sa4ss#92] [Show and tell]
We have all heard the term snake case and camel case; but, did you know there are other cases out there? Today, I just learned about SCREAMING_SNAKE_CASE, which is like snake_case but every letter is upper case. Here is a list of the different cases available and a beautiful infographic from @allison_horst, https://www.allisonhorst.com, showcasing a few of them.
* snake_case
* SCREAMING_SNAKE_CASE or MACRO_CASE
* camelCase, humpCase, or dromedaryCase
* UpperCamelCase or PascalCase
* kebab-case
* Train-Case
* SCREAMING-KEBAB, UPPER-TRAIN-CASE, COBOL-CASE
* flatcase
* dot.case

![image](https://user-images.githubusercontent.com/4108564/230610815-5aedab2c-278e-4b13-96aa-e9fe75f01c3f.png)

There is little to no consistency or requirements for choosing which case to use, especially in R. Java and Kotlin, which have a C and C++ heritage, use camel case for variables and methods and upper camel case for reference types such as enums, classes, and interfaces. In Ruby and Python, SCREAMING_SNAKE_CASE is used for constants.

In R, there are very few hard and fast rules. But, there are some recommendations, largely from [**tidyverse's Style Guide**](https://style.tidyverse.org/), [Advanced R by Hadley Wickham](http://adv-r.had.co.nz/Style.html), and [Google’s R Style Guide](https://google.github.io/styleguide/Rguide.html), where the latter is now based largely on the tidyverse guide itself.
* function names and function argument names should be snake_case
* R markdown chunk names should use kebab-case
* variable names should be nouns
* function names should be verbs

Conclusions for programming in R within [pfmc-assessments](https://github.com/pfmc-assessments)
* individuals should choose a naming convention and stick to it, I recommend tidyverse style but it is not required and 👮 will not come after you if you do not follow my recommendation 🤣 as @iantaylor-NOAA notes below 
* use informative names
* ❌ never use a full stop ( . ), except when defining a new method for a class like S3 or S4
* avoid using names of existing functions and variables, e.g., do not name an R object `plot` because base R has a function already named `plot()`

**Comment by iantaylor-NOAA** [comment]
Can we just go with the tidyverse style? I just look at Google's R Style Guide and would be on board with those few differences from tidyverse except perhaps camel case for functions.

If your point was to choose a naming convention within any given package/repository, where the convention choice would be up to the lead author/maintainer, that's fine with me too.

**Comment by kellijohnson-NOAA** [comment]
Sorry, my comment "choose a naming convention and stick to it" was meant more for individuals rather than a convention as a team. I just created this post to bring awareness about some of the different options that are out there, I had never seen a definition for SCREAMING_SNAKE_CASE before. The bullets under "Conclusions" are things that I plan on trying to adopt in all code that I write. I am a known violator of the last two bullets in many instances ❌ 🤕

**Comment by iantaylor-NOAA** [comment]
I got a big kick out of the Allison Horst graphic and also had never heard of SCREAMING_SNAKE_CASE (which hurts my eyes).
I appreciate your interest in allowing us all flexibility to choose what we want, but if somebody chooses something significantly different from tidyverse, I hope they have a good reason for doing so. 

As a team, we have made some efforts in the past to come up with a coding style, but I'm fine not trying to actually reach consensus on anything because it's not like the style police are going to get called if one of us doesn't follow what we agree on.
* https://docs.google.com/document/d/1CEbHYI2OxqQTfYkrCdB2Maq6aSNDNEbcYRE1dkhXHvQ/edit# (doc from 2014 where we were looking at the Google R Style Guide, before tidyverse style existed)
* The desire to create "Globally accepted coding guidelines for style..." should come out in Fisheries Research in 2023 that may provide some guidance

**Reply by kellijohnson-NOAA** [reply]
@iantaylor-NOAA I edited the second bullet of your post to mention/give a shout out to @chantelwetzel-noaa's paper that should come out in 2023 rather than the outdated idea board / Google doc that Chantel and I were working on prior to this assessment cycle. I hope you don't mind.

**Reply by iantaylor-NOAA** [reply]
No problem. You should probably also add a note to the Google Doc which we all still have access to.
If there's a draft paper that can be shared with the team, that would be welcome too.

**Reply by kellijohnson-NOAA** [reply]
I will try to remember to share the draft once it comes back from review. Spoiler alert ... use "spawning biomass" rather than "spawning stock biomass" in your text 🥇.

**Reply by iantaylor-NOAA** [reply]
Sounds good. Any spoiler alert for fraction unfished/% unfished/relative spawning biomass/depletion/biomass ratio?

**Reply by brianlangseth-NOAA** [reply]
Once it does come out could we add some of the language conventions into the handbook? Would be great to incorporate into this cycles assessments but that may be one task too many as folks try to write up.
